add std.conv.hexString() to replace core language hex strings

WalterBright opened the pull request:
This is a reboot of https://github.com/D-Programming-Language/phobos/pull/3058 written by @BBasile which was mysteriously closed.

Forum discussion: http://forum.dlang.org/thread/me4njq%241cul%241@digitalmars.com

JakobOvrum commented:
I mentioned a number of issues in the other PR that Github now classifies as "outdated diffs" but the issues were not actually addressed. Can you have a look at them?

WalterBright commented:
Links to which specific ones?

JakobOvrum commented:
* `static` is redundant for module-level functions
 * s/`hexStrImpl!(C)(hexData)`/`hexStrImpl!C(hexData)`/ etc
 * user-facing compile-time interface should be value templates, not function templates:
```d
template hexString(string hexData, C = char)
    if (hexData.isHexLiteral && isSomeChar!C)
{
    immutable hexString = hexStrImpl!C(hexData);
}
```

------------

My main point was that I don't think we should expose the runtime functions as they are proposed by this patch, as they're not written as range algorithms.

Panke commented on std/conv.d:
```diff
@@ -5334,3 +5334,249 @@ unittest
+/**
+The function $(D isHexLiteral) checks the consistency of a string for the
+functions $(D hexString) and $(D hexBytes).
+The result is only true if the input string is composed of ASCII white
```
nitpick:  only true if ==> true if and only if

Panke commented on std/conv.d:
```diff
@@ -5334,3 +5334,249 @@ unittest
+        i += isH;
+    }
+    return !(i & 1) & (i > 0);
+}
```
Use logical operators for logical operations and bit operators for bit operations.

Panke commented on std/conv.d:
```diff
@@ -5334,3 +5334,249 @@ unittest
+    static assert( ("0123456789abcdefABCDEF").isHexLiteral);
+    // empty or white strings are not valid
+    static assert( !"\r\n\t".isHexLiteral);
+    // but are accepted if the count of hex digits is even
```
Why? At least mention that in the docs.

JakobOvrum commented:
Ping @WalterBright

WalterBright commented:
rebased

WalterBright commented:
Revamped it according to comments here. Removed hexBytes. Adjusted semantics to exactly match hex strings in the core language, except that nelSep is also recognized as whitespace.

This is meant to replace core hex strings, not as something more. Something more may come about in the future as usage leads us.

JakobOvrum commented:
Nice! LGTM.

kuettler commented on std/conv.d:
```diff
@@ -5350,3 +5350,195 @@ unittest
+{
+    // test all the hex digits
+    static assert( ("0123456789abcdefABCDEF").isHexLiteral);
+    // empty or white strings are not valid
```
Is there one not too many in the comment?

kuettler commented on std/conv.d:
```diff
@@ -5350,3 +5350,195 @@ unittest
+    static assert( ("0123456789abcdefABCDEF").isHexLiteral);
+    // empty or white strings are not valid
+    static assert( "\r\n\t".isHexLiteral);
+    // but are accepted if the count of hex digits is even
```
Again, the comment is misleading

kuettler commented:
LGTM

MartinNowak commented:
Auto-merge toggled on
